pkg/cmd/roachtest/tests/idempotent_permanent_upgrades.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package tests
+
+import (
+	"context"
+	"math/rand"
+
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/registry"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/roachtestutil/mixedversion"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/install"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/logger"
+	"github.com/cockroachdb/errors"
+)
+
+func registerIdempotentPermanentUpgrades(r registry.Registry) {
+	r.Add(registry.TestSpec{
+		Name:             "upgrade/idempotent-permanent-upgrades",
+		Owner:            registry.OwnerServer,
+		Cluster:          r.MakeClusterSpec(1),
+		CompatibleClouds: registry.AllExceptAWS,
+		Suites:           registry.Suites(registry.Nightly),
+		Randomized:       false,
+		Run: func(ctx context.Context, t test.Test, c cluster.Cluster) {
+			runIdempotentPermanentUpgrades(ctx, t, c)
+		},
+	})
+}
+
+func runIdempotentPermanentUpgrades(ctx context.Context, t test.Test, c cluster.Cluster) {
+	// This test isn't really a "mixed version test" per se, but it uses the
+	// mixedversion framework since it makes it very easy to test upgrades from
+	// older versions.
+	mvt := mixedversion.NewTest(
+		ctx, t, t.L(), c, c.All(),
+		// We want to start from v22.1 or earlier.
+		mixedversion.MinUpgrades(4),
```

Comment:
I wonder if should add `mixedversion.MustUpgrades(x, y)` (to the framework) since `MinUpgrades` will rot over time, i.e., as of v25.2.

Reply:
Assuming it's the first such use-case, where an absolute upgrade path must be specified, it'll be updated when [1] is tackled.

[1] https://github.com/cockroachdb/cockroach/issues/136223

Reply:
@srosenberg thesis being merged specifically to 24.1, not master, so rotting in a future release doesn't seem like a serious concern here does it?